src/map_ref.rs

```diff
@@ -0,0 +1,272 @@
+use crate::iter::*;
+use crate::HashMap;
+use crossbeam_epoch::Guard;
+use std::borrow::Borrow;
+use std::fmt::{self, Debug, Formatter};
+use std::hash::{BuildHasher, Hash};
+use std::ops::{Deref, Index};
+
+/// A reference to a [`HashMap`], constructed with [`HashMap::guarded`] or [`HashMap::with_guard`].
+///
+/// The current thread will be pinned for the duration of this reference.
+/// Keep in mind that this prevents the collection of garbage generated by the map.
+pub struct HashMapRef<'map, K: 'static, V: 'static, S = crate::DefaultHashBuilder> {
+    map: &'map HashMap<K, V, S>,
+    guard: GuardRef<'map>,
+}
+
+enum GuardRef<'g> {
+    Owned(Guard),
+    Ref(&'g Guard),
+}
+
+impl Deref for GuardRef<'_> {
+    type Target = Guard;
+
+    #[inline]
+    fn deref(&self) -> &Guard {
+        match *self {
+            GuardRef::Owned(ref guard) | GuardRef::Ref(&ref guard) => guard,
+        }
+    }
+}
+
+impl<K, V, S> HashMap<K, V, S>
+where
+    K: Sync + Send + Clone + Hash + Eq,
+    V: Sync + Send,
+    S: BuildHasher,
+{
+    /// Get a reference to this map with the current thread pinned.
+    ///
+    /// Keep in mind that for as long as you hold onto this, you are preventing the collection of
+    /// garbage generated by the map.
+    pub fn guarded(&self) -> HashMapRef<'_, K, V, S> {
```

Comment:
I wonder if we want to call this method `pin` to indicate that that's _really_ what you're doing.

Reply:
I was actually a little surprised on `fn guard(&self) -> Guard` that you didn't call that `pin`, since it ends with calling `pin()` inside. But I'm ok if you want to call this `pin` instead.

Reply:
So I actually didn't call it `pin` specifically because I wanted _this_ method to be called `pin`. My thinking here is that it'll often make _more_ sense to `pin` and then operate on a `HashMapRef` than to manually get a `Guard` and then pass that to a bunch of methods. Making the `pin` method return a `HashMapRef` makes it more likely that people will reach for and discover this way of using the map, which I think we want to encourage.